File: C#/WindowsPhone/LGame-XNA-lib/Loon.Utils.Collection/IntArray.cs
using System;
namespace Loon.Utils.Collection
{
    
    public class IntArray 
    {

        public int[] items;
        public int size;
        public bool ordered;

        public IntArray():this(true, CollectionUtils.INITIAL_CAPACITY)
        {
            
        }

        public IntArray(string ctx)
        {
            string intString = ctx;
            if (intString != null)
            {
                if (intString.IndexOf(',') != -1)
                {
                    intString = StringUtils.Replace(intString, ",", "")
                            .Replace(" ", "").Trim();
                }
                if (!MathUtils.IsNan(intString))
                {
                    throw new Exception('[' + intString + ']'
                            + " not number ! ");
                }
            }
            else
            {
                throw new Exception('[' + intString + ']' + " is NULL ! ");
            }
            char[] tmp = intString.ToCharArray();
            int len = tmp.Length;
            items = new int[len];
            for (int i = 0; i < len; i++)
            {
                switch ((int)tmp[i])
                {
                    case '0':
                        items[i] = 0;
                        break;
                    case '1':
                        items[i] = 1;
                        break;
                    case '2':
                        items[i] = 2;
                        break;
                    case '3':
                        items[i] = 3;
                        break;
                    case '4':
                        items[i] = 4;
                        break;
                    case '5':
                        items[i] = 5;
                        break;
                    case '6':
                        items[i] = 6;
                        break;
                    case '7':
                        items[i] = 7;
                        break;
                    case '8':
                        items[i] = 8;
                        break;
                    case '9':
                        items[i] = 9;
                        break;
                    default:
                        items[i] = -1;
                        break;
                }
            }
            size = len;
            tmp = null;
        }

        public IntArray(int capacity):this(true, capacity)
        {
            
        }

        public IntArray(bool o, int capacity)
        {
            this.ordered = o;
            items = new int[capacity];
        }

        public IntArray(IntArray array)
        {
            this.ordered = array.ordered;
            size = array.size;
            items = new int[size];
            System.Array.Copy(array.items, 0, items, 0, size);
        }

        public IntArray(params int[] array): this(true, array)
        {
           
        }

        public IntArray(bool o, int[] array): this(o, array.Length)
        {
            size = array.Length;
            System.Array.Copy(array, 0, items, 0, size);
        }

        public void Add(int value_ren)
        {
            int[] item = this.items;
            if (size == item.Length)
            {
                item = Resize(MathUtils.Max(8, (int)(size * 1.75f)));
            }
            item[size++] = value_ren;
        }

        public void AddAll(IntArray array)
        {
            AddAll(array, 0, array.size);
        }

        public void AddAll(IntArray array, int offset, int length)
        {
            if (offset + length > array.size)
            {
                throw new ArgumentException(
                        "offset + length must be <= size: " + offset + " + "
                                + length + " <= " + array.size);
            }
            AddAll(array.items, offset, length);
        }

        public void AddAll(int[] array)
        {
            AddAll(array, 0, array.Length);
        }

        public void AddAll(int[] array, int offset, int length)
        {
            int[] item = this.items;
            int sizeNeeded = size + length - offset;
            if (sizeNeeded >= item.Length)
            {
                item = Resize(MathUtils.Max(8, (int)(sizeNeeded * 1.75f)));
            }
            System.Array.Copy((Array)(array), offset, (Array)(item), size, length);
            size += length;
        }

        public int Get(int index)
        {
            if (index >= size)
            {
                throw new IndexOutOfRangeException(index.ToString().ToString());
            }
            return items[index];
        }

        public void Set(int index, int value_ren)
        {
            if (index >= size)
            {
                throw new IndexOutOfRangeException(index.ToString().ToString());
            }
            items[index] = value_ren;
        }

        public void Insert(int index, int value_ren)
        {
            int[] item = this.items;
            if (size == item.Length)
                item = Resize(Math.Max(8, (int)(size * 1.75f)));
            if (ordered)
            {
                System.Array.Copy((Array)(item), index, (Array)(item), index + 1, size - index);
            }
            else
            {
                item[size] = item[index];
            }
            size++;
            item[index] = value_ren;
        }

        public bool Contains(int value_ren)
        {
            int i = size - 1;
            int[] item = this.items;
            while (i >= 0)
            {
                if (item[i--] == value_ren)
                    return true;
            }
            return false;
        }

        public bool ContainsAll(IntArray c)
        {
            for (int i = 0; i < c.size; i++)
            {
                if (!Contains(c.items[i]))
                {
                    return false;
                }
            }
            return true;
        }

        public bool Part(IntArray c)
        {
            int cc = 0;
            for (int i = 0; i < c.size; i++)
            {
                if (Contains(c.items[i]))
                {
                    cc++;
                }
            }
            return cc != 0;
        }

        public int IndexOf(int value_ren)
        {
            int[] item = this.items;
            for (int i = 0, n = size; i < n; i++)
            {
                if (item[i] == value_ren)
                {
                    return i;
                }
            }
            return -1;
        }

        public int Count(int value_ren)
        {
            int cc = 0;
            int[] item = this.items;
            for (int i = 0; i < size; i++)
            {
                if (item[i] == value_ren)
                {
                    cc++;
                }
            }
            return cc;
        }

        public IntArray Copy()
        {
            return new IntArray(this);
        }

        public bool RetainAll(IntArray c)
        {
            return BatchRemove(c, true);
        }

        public bool RemoveAll(IntArray c)
        {
            return BatchRemove(c, false);
        }

        private bool BatchRemove(IntArray c, bool complement)
        {
            int[] data = this.items;
            int r = 0, w = 0;
            bool modified = false;
            try
            {
                for (; r < size; r++)
                {
                    if (c.Contains(data[r]) == complement)
                    {
                        data[w++] = data[r];
                    }
                }
            }
            finally
            {
                if (r != size)
                {
                    System.Array.Copy((Array)(data), r, (Array)(data), w, size - r);
                    w += size - r;
                }
                if (w != size)
                {
                    for (int i = w; i < size; i++)
                    {
                        data[i] = -1;
                    }
                    size = w;
                    modified = true;
                }
            }
            return modified;
        }

        public bool RemoveValue(int value_ren)
        {
            int[] item = this.items;
            for (int i = 0, n = size; i < n; i++)
            {
                if (item[i] == value_ren)
                {
                    RemoveIndex(i);
                    return true;
                }
            }
            return false;
        }

        public int RemoveIndex(int index)
        {
            if (index >= size)
            {
                throw new IndexOutOfRangeException(index.ToString().ToString());
            }
            int[] item = this.items;
            int value_ren = item[index];
            size--;
            if (ordered)
            {
                System.Array.Copy(item, index + 1,item, index, size - index);
            }
            else
            {
                item[index] = item[size];
            }
            return value_ren;
        }

        public int Pop()
        {
            return items[--size];
        }

        public int Peek()
        {
            return items[size - 1];
        }

        public void Clear()
        {
            size = 0;
        }

        public int[] EnsureCapacity(int additionalCapacity)
        {
            int sizeNeeded = size + additionalCapacity;
            if (sizeNeeded >= items.Length)
            {
                Resize(Math.Max(8, sizeNeeded));
            }
            return items;
        }

        protected internal int[] Resize(int newSize)
        {
            int[] newItems = new int[newSize];
            int[] item = this.items;
            System.Array.Copy(item, 0,newItems, 0, Math.Min(item.Length, newItems.Length));
            this.items = newItems;
            return newItems;
        }

        public void Sort()
        {
            System.Array.Sort(items, 0, size);
        }

        public int[] ToArray()
        {
            int[] array = new int[size];
            System.Array.Copy((Array)(items), 0, (Array)(array), 0, size);
            return array;
        }

        public string ToString(char split)
        {
            if (size == 0)
            {
                return "[]";
            }
            int[] item = this.items;
            System.Text.StringBuilder buffer = new System.Text.StringBuilder(
                    CollectionUtils.INITIAL_CAPACITY);
            buffer.Append('[');
            buffer.Append(item[0]);
            for (int i = 1; i < size; i++)
            {
                buffer.Append(split);
                buffer.Append(item[i]);
            }
            buffer.Append(']');
            return buffer.ToString();
        }

        public override string ToString()
        {
            return ToString(',');
        }
    }
}

File: Loon-0.5-Alpha/gwt/src/loon/utils/collection/IntArray.java
/**
 * Copyright 2014
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.4.2
 */
package loon.utils.collection;

import java.math.BigInteger;
import java.util.Arrays;

import loon.utils.CollectionUtils;
import loon.utils.MathUtils;

public class IntArray {
	public int[] items;
	public int length;
	public boolean ordered;

	public IntArray() {
		this(true, 16);
	}

	public IntArray(int capacity) {
		this(true, capacity);
	}

	public IntArray(boolean ordered, int capacity) {
		this.ordered = ordered;
		items = new int[capacity];
	}

	public IntArray(IntArray array) {
		this.ordered = array.ordered;
		length = array.length;
		items = new int[length];
		System.arraycopy(array.items, 0, items, 0, length);
	}

	public IntArray(int[] array) {
		this(true, array, 0, array.length);
	}

	public IntArray(boolean ordered, int[] array, int startIndex, int count) {
		this(ordered, count);
		length = count;
		System.arraycopy(array, startIndex, items, 0, count);
	}

	public void unshift(int value) {
		if (length > 0) {
			int[] items = this.items;
			int[] newItems = new int[length + 1];
			newItems[0] = value;
			System.arraycopy(items, 0, newItems, 1, length);
			this.length = newItems.length;
			this.items = newItems;
		} else {
			add(value);
		}
	}

	public void push(int value) {
		add(value);
	}

	public void add(int value) {
		int[] items = this.items;
		if (length == items.length) {
			items = relength(Math.max(8, (int) (length * 1.75f)));
		}
		items[length++] = value;
	}

	public void addAll(IntArray array) {
		addAll(array, 0, array.length);
	}

	public void addAll(IntArray array, int offset, int length) {
		if (offset + length > array.length)
			throw new IllegalArgumentException(
					"offset + length must be <= length: " + offset + " + "
							+ length + " <= " + array.length);
		addAll(array.items, offset, length);
	}

	public void addAll(int... array) {
		addAll(array, 0, array.length);
	}

	public void addAll(int[] array, int offset, int length) {
		int[] items = this.items;
		int lengthNeeded = length + length;
		if (lengthNeeded > items.length) {
			items = relength(Math.max(8, (int) (lengthNeeded * 1.75f)));
		}
		System.arraycopy(array, offset, items, length, length);
		length += length;
	}

	public int get(int index) {
		if (index >= length) {
			return 0;
		}
		return items[index];
	}

	public void set(int index, int value) {
		if (index >= length) {
			int size = length;
			for (int i = size; i < index + 1; i++) {
				add(0);
			}
			items[index] = value;
			return;
		}
		items[index] = value;
	}

	public void incr(int index, int value) {
		if (index >= length)
			throw new IndexOutOfBoundsException("index can't be >= length: "
					+ index + " >= " + length);
		items[index] += value;
	}

	public void mul(int index, int value) {
		if (index >= length)
			throw new IndexOutOfBoundsException("index can't be >= length: "
					+ index + " >= " + length);
		items[index] *= value;
	}

	public void insert(int index, int value) {
		if (index > length) {
			throw new IndexOutOfBoundsException("index can't be > length: "
					+ index + " > " + length);
		}
		int[] items = this.items;
		if (length == items.length)
			items = relength(Math.max(8, (int) (length * 1.75f)));
		if (ordered)
			System.arraycopy(items, index, items, index + 1, length - index);
		else
			items[length] = items[index];
		length++;
		items[index] = value;
	}

	public void swap(int first, int second) {
		if (first >= length)
			throw new IndexOutOfBoundsException("first can't be >= length: "
					+ first + " >= " + length);
		if (second >= length)
			throw new IndexOutOfBoundsException("second can't be >= length: "
					+ second + " >= " + length);
		int[] items = this.items;
		int firstValue = items[first];
		items[first] = items[second];
		items[second] = firstValue;
	}

	public boolean contains(int value) {
		int i = length - 1;
		int[] items = this.items;
		while (i >= 0)
			if (items[i--] == value)
				return true;
		return false;
	}

	public int indexOf(int value) {
		int[] items = this.items;
		for (int i = 0, n = length; i < n; i++)
			if (items[i] == value)
				return i;
		return -1;
	}

	public int lastIndexOf(int value) {
		int[] items = this.items;
		for (int i = length - 1; i >= 0; i--)
			if (items[i] == value)
				return i;
		return -1;
	}

	public boolean removeValue(int value) {
		int[] items = this.items;
		for (int i = 0, n = length; i < n; i++) {
			if (items[i] == value) {
				removeIndex(i);
				return true;
			}
		}
		return false;
	}

	public int removeIndex(int index) {
		if (index >= length) {
			throw new IndexOutOfBoundsException("index can't be >= length: "
					+ index + " >= " + length);
		}
		int[] items = this.items;
		int value = items[index];
		length--;
		if (ordered) {
			System.arraycopy(items, index + 1, items, index, length - index);
		} else {
			items[index] = items[length];
		}
		return value;
	}

	public void removeRange(int start, int end) {
		if (end >= length) {
			throw new IndexOutOfBoundsException("end can't be >= length: "
					+ end + " >= " + length);
		}
		if (start > end) {
			throw new IndexOutOfBoundsException("start can't be > end: "
					+ start + " > " + end);
		}
		int[] items = this.items;
		int count = end - start + 1;
		if (ordered) {
			System.arraycopy(items, start + count, items, start, length
					- (start + count));
		} else {
			int lastIndex = this.length - 1;
			for (int i = 0; i < count; i++)
				items[start + i] = items[lastIndex - i];
		}
		length -= count;
	}

	public boolean removeAll(IntArray array) {
		int length = this.length;
		int startlength = length;
		int[] items = this.items;
		for (int i = 0, n = array.length; i < n; i++) {
			int item = array.get(i);
			for (int ii = 0; ii < length; ii++) {
				if (item == items[ii]) {
					removeIndex(ii);
					length--;
					break;
				}
			}
		}
		return length != startlength;
	}

	public int pop() {
		return items[--length];
	}

	public int shift() {
		return removeIndex(0);
	}

	public int peek() {
		return items[length - 1];
	}

	public int first() {
		if (length == 0) {
			throw new IllegalStateException("Array is empty.");
		}
		return items[0];
	}

	public void clear() {
		length = 0;
	}

	public int[] shrink() {
		if (items.length != length)
			relength(length);
		return items;
	}

	public int[] ensureCapacity(int additionalCapacity) {
		int lengthNeeded = length + additionalCapacity;
		if (lengthNeeded > items.length)
			relength(Math.max(8, lengthNeeded));
		return items;
	}

	protected int[] relength(int newlength) {
		int[] newItems = new int[newlength];
		int[] items = this.items;
		System.arraycopy(items, 0, newItems, 0,
				Math.min(length, newItems.length));
		this.items = newItems;
		return newItems;
	}

	public void sort() {
		Arrays.sort(items, 0, length);
	}

	public void reverse() {
		int[] items = this.items;
		for (int i = 0, lastIndex = length - 1, n = length / 2; i < n; i++) {
			int ii = lastIndex - i;
			int temp = items[i];
			items[i] = items[ii];
			items[ii] = temp;
		}
	}

	public void shuffle() {
		int[] items = this.items;
		for (int i = length - 1; i >= 0; i--) {
			int ii = MathUtils.random(i);
			int temp = items[i];
			items[i] = items[ii];
			items[ii] = temp;
		}
	}

	public void truncate(int newlength) {
		if (length > newlength)
			length = newlength;
	}

	public int random() {
		if (length == 0) {
			return 0;
		}
		return items[MathUtils.random(0, length - 1)];
	}

	public int[] toArray() {
		int[] array = new int[length];
		System.arraycopy(items, 0, array, 0, length);
		return array;
	}

	public boolean equals(Object object) {
		if (object == this)
			return true;
		if (!(object instanceof IntArray))
			return false;
		IntArray array = (IntArray) object;
		int n = length;
		if (n != array.length)
			return false;
		for (int i = 0; i < n; i++)
			if (items[i] != array.items[i])
				return false;
		return true;
	}

	public String toString(String separator) {
		if (length == 0)
			return "";
		int[] items = this.items;
		StringBuilder buffer = new StringBuilder(32);
		buffer.append(items[0]);
		for (int i = 1; i < length; i++) {
			buffer.append(separator);
			buffer.append(items[i]);
		}
		return buffer.toString();
	}

	static public IntArray with(int... array) {
		return new IntArray(array);
	}

	public IntArray splice(int begin, int end) {
		IntArray longs = new IntArray(slice(begin, end));
		if (end - begin >= length) {
			items = new int[0];
			length = 0;
			return longs;
		} else {
			removeRange(begin, end - 1);
		}
		return longs;
	}

	public static int[] slice(int[] array, int begin, int end) {
		if (begin > end) {
			throw new RuntimeException();
		}
		if (begin < 0) {
			begin = array.length + begin;
		}
		if (end < 0) {
			end = array.length + end;
		}
		int elements = end - begin;
		int[] ret = new int[elements];
		System.arraycopy(array, begin, ret, 0, elements);
		return ret;
	}

	public static int[] slice(int[] array, int begin) {
		return slice(array, begin, array.length);
	}

	public IntArray slice(int size) {
		return new IntArray(slice(this.items, size, this.length));
	}

	public IntArray slice(int begin, int end) {
		return new IntArray(slice(this.items, begin, end));
	}

	public static int[] concat(int[] array, int[] other) {
		return concat(array, array.length, other, other.length);
	}

	public static int[] concat(int[] array, int alen, int[] other, int blen) {
		int[] ret = new int[alen + blen];
		System.arraycopy(array, 0, ret, 0, alen);
		System.arraycopy(other, 0, ret, alen, blen);
		return ret;
	}

	public IntArray concat(IntArray o) {
		return new IntArray(concat(this.items, this.length, o.items, o.length));
	}

	public byte[] getBytes() {
		int[] items = this.items;
		byte[] bytes = new byte[this.length];
		for (int i = 0; i < bytes.length; i++) {
			bytes[i] = BigInteger.valueOf(items[i]).byteValue();
		}
		return bytes;
	}

	public String toString(char split) {
		if (length == 0) {
			return "[]";
		}
		int[] items = this.items;
		StringBuilder buffer = new StringBuilder(
				CollectionUtils.INITIAL_CAPACITY);
		buffer.append('[');
		buffer.append(items[0]);
		for (int i = 1; i < length; i++) {
			buffer.append(split);
			buffer.append(items[i]);
		}
		buffer.append(']');
		return buffer.toString();
	}

	@Override
	public String toString() {
		return toString(',');
	}
}
